### ISSUE TITLE
Questionable results with prepare

### ISSUE THREAD

**Issue by pgulutzan** [tarantool/tarantool#5945]
I have Ubuntu 20.04. I have Tarantool 2.7, pulled from source today.

I execute these requests:
```lua
box.execute([[CREATE TABLE n (s1 INT PRIMARY KEY, s2 INT DEFAULT 5);]])
box.execute([[INSERT INTO n VALUES (1,NULL);]])
prep = box.prepare([[SELECT * FROM (SELECT * FROM n) AS x ORDER BY 2 LIMIT 5;]])
box.execute([[TRUNCATE TABLE n;]])
prep:execute()
```
Result = `Aborted (core dumped)`
In fact, I'm not sure whether the result should be zero rows or an
error = `statement has expired`. Either way is okay.

I execute these requests:
```lua
box.execute([[CREATE TABLE o (s1 INT PRIMARY KEY, s2 INT DEFAULT 5);]])
box.execute([[INSERT INTO o VALUES (1,NULL), (2,-1),(3,-2);]])
prep = box.prepare([[SELECT * FROM o ORDER BY :b,:a;]])
prep:execute({{[':a'] = 1}, {[':b'] = 2}})
```
Result = rows in order by column 1
In fact, I think this is correct, even though
```lua
box.execute([[SELECT * FROM o ORDER BY 2,1;]])
```
would return rows in order by column 2.
I am only looking for confirmation: this is expected behaviour, right?

I execute these requests:
```lua
box.execute([[DROP TABLE q;]])
box.space._session_settings:update('sql_default_engine', {{'=', 'value', 'vinyl'}})
prep = box.prepare([[CREATE TABLE q (s1 INT PRIMARY KEY);]])
box.space._session_settings:update('sql_default_engine', {{'=', 'value', 'memtx'}})
prep:execute()
box.space._space:select()
```
Result = Q has a vinyl engine.
I guess that means: it is decided at the time of the prepare, not at the time of the execute.
Again, I am only looking for confirmation: this is expected behaviour, right?

**Comment by AyushKadwe** [comment]
Hi, I want work on this issue , I already work on SQL and PLSQL database, but this is my first time to contribute in open source . Can you guide me on this issue?

**Comment by kyukhin** [comment]
> I have Ubuntu 20.04. I have Tarantool 2.7, pulled from source today.
> 
> I execute these requests:
> 
> ```lua
> box.execute([[CREATE TABLE n (s1 INT PRIMARY KEY, s2 INT DEFAULT 5);]])
> box.execute([[INSERT INTO n VALUES (1,NULL);]])
> prep = box.prepare([[SELECT * FROM (SELECT * FROM n) AS x ORDER BY 2 LIMIT 5;]])
> box.execute([[TRUNCATE TABLE n;]])
> prep:execute()
> ```
> 
> Result = `Aborted (core dumped)`
> In fact, I'm not sure whether the result should be zero rows or an
> error = `statement has expired`. Either way is okay.

Bug indeed.

> I execute these requests:
> 
> ```lua
> box.execute([[CREATE TABLE o (s1 INT PRIMARY KEY, s2 INT DEFAULT 5);]])
> box.execute([[INSERT INTO o VALUES (1,NULL), (2,-1),(3,-2);]])
> prep = box.prepare([[SELECT * FROM o ORDER BY :b,:a;]])
> prep:execute({{[':a'] = 1}, {[':b'] = 2}})
> ```
> 
> Result = rows in order by column 1
> In fact, I think this is correct, even though
> 
> ```lua
> box.execute([[SELECT * FROM o ORDER BY 2,1;]])
> ```
> 
> would return rows in order by column 2.
> I am only looking for confirmation: this is expected behaviour, right?

I guess this should be documented.

> I execute these requests:
> 
> ```lua
> box.execute([[DROP TABLE q;]])
> box.space._session_settings:update('sql_default_engine', {{'=', 'value', 'vinyl'}})
> prep = box.prepare([[CREATE TABLE q (s1 INT PRIMARY KEY);]])
> box.space._session_settings:update('sql_default_engine', {{'=', 'value', 'memtx'}})
> prep:execute()
> box.space._space:select()
> ```
> 
> Result = Q has a vinyl engine.
> I guess that means: it is decided at the time of the prepare, not at the time of the execute.
> Again, I am only looking for confirmation: this is expected behaviour, right?

I don't see the problem here. You created a space on top of the vinyl engine. Why engine should be changed depending on session settings?

**Comment by pgulutzan** [comment]
@kyukhin:
Re your reply "I guess this should be documented":
Yes -- if it is expected behaviour.
The documentation change would be:
in https://www.tarantool.io/en/doc/latest/reference/reference_sql/sql_statements_and_clauses/#order-by-clause
I could change "Expression is a positive integer" to "Expression is a positive integer literal".
Re your reply "I do not see the problem here".
There isn't a problem -- if it is expected behaviour.
I did not "create a space on top of the vinyl engine", I merely prepared to create a space,
I actually created the space after the setting was changed to memtx.
